### PR TITLE
timestamps

### PR DESCRIPTION
**Pull request by pwoody**

### CODE REVIEW

**Comment by pwoody** [comment]
@robert3005 fixed the weird rebase. this should be good to go

**Comment by pwoody** [comment]
This works on the legacy int96 format as well because `spark.sql.parquet.int96AsTimestamp` defaults to true and disables filter pushdown. If you set that setting to false you will see this error when attempting to read the old format:

`Caused by: java.lang.IllegalArgumentException: FilterPredicate column: value's declared type (java.lang.Long) does not match the schema found in file metadata. Column value is of type: INT96
Valid types for this column are: [class org.apache.parquet.io.api.Binary]`

**Comment by pwoody** [comment]
@ash211 @lynchmattj @schlosna

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 1]
```diff
@@ -397,12 +401,12 @@ class ParquetFileFormat
...
           // converted (`ParquetFilters.createFilter` returns an `Option`). That's why a `flatMap`
           // is used here.
-          .flatMap(ParquetFilters.createFilter(requiredSchema, _))
+          .flatMap(ParquetFilters.createFilter(requiredSchema, _, int96AsTimestamp))
           .reduceOption(FilterApi.and)
       } else {
         None
       }
-
+    log.debug(s"Pushing converted filters: $pushed")
```
Comment (robert3005) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: maybe info? FilterCompat.get logs at that level

**Comment by robert3005** [comment]
👍 only nit about logging since I was chasing about that value multiple times already. Otherwise looks good to me.

**Comment by pwoody** [comment]
Cool done, merging.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 2]
```diff
@@ -153,23 +206,32 @@ private[parquet] object ParquetFilters {
...
+        // root fields. Currently, accessing to nested fields does not push down filters
+        // and it does not support to create filters for them.
+        // scalastyle:off println
+        fields.filterNot { f =>
+          val isTs = DataTypes.TimestampType.acceptsType(f.dataType)
+
+          val isOptionalField = f.metadata.contains(StructType.metadataKeyForOptionalField) &&
+            f.metadata.getBoolean(StructType.metadataKeyForOptionalField)
+
+          (isTs && int96AsTimestamp) || isOptionalField
```
Comment (schlosna) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: nit: could be lazy with the boolean evaluations:

```
fields.filterNot { f =>
    (int96AsTimestamp && DataTypes.TimestampType.acceptsType(f.dataType))
    || (f.metadata.contains(StructType.metadataKeyForOptionalField)
                    && f.metadata.getBoolean(StructType.metadataKeyForOptionalField))
}.map(f => f.name -> f.dataType).toMap
```

**Pull request merged**